### DISCUSSION TITLE
datasets/detect/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#12639] [Docs]
# datasets/detect/

Navigate through supported dataset formats, methods to utilize them and how to add your own datasets. Get insights on porting or converting label formats.

https://docs.ultralytics.com/datasets/detect/

**Comment by nepulepu** [comment]
Hello,
I have exported yolov8 dataset on roboflow but the format of the annotation seems off,
the annotation files looks like this:
``` 
0 0.4505104166666667 0.054072916666666665 0.4505104166666667 0.004177083333333333 0.3871354166666666 0.004177083333333333 0.3871354166666666 0.054072916666666665 0.4505104166666667 0.054072916666666665
0 0.48017708333333337 0.39927083333333335 0.48017708333333337 0.34398958333333335 0.43028125 0.34398958333333335 0.43028125 0.39927083333333335 0.48017708333333337 0.39927083333333335
0 0.5647291666666666 0.05663541666666667 0.5647291666666666 0.009708333333333334 0.50646875 0.009708333333333334 0.50646875 0.05663541666666667 0.5647291666666666 0.05663541666666667
0 0.44916666666666666 0 0.44916666666666666 0.028447916666666666 0.5111875 0.028447916666666666 0.5111875 0 0.44916666666666666 0
0 0.11178125 0.5493645833333333 0.11178125 0.49411458333333336 0.05559375 0.49411458333333336 0.05559375 0.5493645833333333 0.11178125 0.5493645833333333
0 0.09740625 0.18059375 0.09740625 0.11347916666666666 0.04121875 0.11347916666666666 0.04121875 0.18059375 0.09740625 0.18059375
0 0 0.0935 0.0285 0.0935 0.0285 0.020927083333333332 0 0.020927083333333332 0 0.0935
0 0.40507291666666667 0.2566041666666667 0.40507291666666667 0.21210416666666668 0.35517708333333337 0.21210416666666668 0.35517708333333337 0.2566041666666667 0.40507291666666667 0.2566041666666667
0 0.5711875000000001 0.6439687500000001 0.5711875000000001 0.5838125000000001 0.5115833333333334 0.5838125000000001 0.5115833333333334 0.6439687500000001 0.5711875000000001 0.6439687500000001
```

as far as I know the yolo format is : (class, center-x, center-y, width, height) and the annotation above has more than 5 points, when training the model it seems fine but what are the excessive points represent? does it affect model perfomance?

**Reply by pderrenger** [reply]
Hello,

It looks like the annotation you're seeing is more complex than the standard YOLO format because it includes coordinates for oriented bounding boxes (OBB), which are used to describe rotated rectangles. The standard YOLO format is indeed `(class, center-x, center-y, width, height)`, but it seems your data might have been prepared with additional points for more detailed annotations, especially in scenarios where orientation matters, such as satellite imagery or certain types of scientific data.

This additional data shouldn't affect model performance negatively; it’s providing more detailed information that could be very useful if the model and the training setup are designed to utilize OBB. If you're training a regular YOLO model that expects standard bounding boxes, you'll want to convert these annotations or use a model that can interpret rotated bounding boxes.

To get a better understanding and potentially adjust your dataset annotations, you might find the section on porting or converting label formats in the [Ultralytics Docs](https://docs.ultralytics.com/datasets/detect/) helpful.

👍 Happy training!

**Reply by nepulepu** [reply]
Hello,
Thanks for replying, I may know the reason this happens, as I import coco format annotation which has segmentation in the annotation file.
It causes the bbox become polygon, is there any way to export into normal yolo bbox or i need to re-edit my annotation files?
Thanks.

**Reply by pderrenger** [reply]
@nepulepu hello,

It looks like you've encountered an issue with the polygon bounding boxes derived from the segmentation data in your COCO format annotations. For training YOLO models, you'll typically need the annotations in the YOLO format, which expects bounding boxes as `(class, center_x, center_y, width, height)`.

You can convert your polygon annotations to regular bounding boxes by calculating the bounding rectangle that encapsulates your polygon. Here’s a simple way you can modify your existing annotations:

1. For each polygon, find the minimum and maximum x and y coordinates.
2. Use these coordinates to calculate the bounding box’s center, width, and height.

Here’s a short example in Python, assuming you have polygon coordinates:
```python
import numpy as np

# Example polygon coordinates
polygon = np.array([[x1, y1], [x2, y2], [x3, y3], [x4, y4]])

# Calculate bounding box
xmin, ymin = np.min(polygon, axis=0)
xmax, ymax = np.max(polygon, axis=0)
x_center = (xmin + xmax) / 2
y_center = (ymin + ymax) / 2
width = xmax - xmin
height = ymax - ymin

# Normalize by image dimensions if necessary
x_center /= image_width
y_center /= image_height
width /= image_width
height /= image_height

# YOLO format: class_id, x_center, y_center, width, height
yolo_bbox = [class_id, x_center, y_center, width, height]
```

Above, `x1, y1, ... x4, y4` are the coordinates of the polygon’s vertices, and `class_id` is the object class. This would convert your segmentation polygons to the bounding boxes needed for YOLO.

If editing multiple files manually seems laborious, consider scripting this process or using tools designed to convert from COCO to YOLO format automatically, like the one mentioned in the Ultralytics documentation. This way, you can prepare your dataset efficiently for YOLO training.

Good luck with your model training! 👍

**Reply by skp4595** [reply]
just a follow up question here

I converted the dataset accordingly (images and it's labels), now I have used this class_mapping = {'Car': 0, 'Van': 1, 'Truck': 2, 'Pedestrian': 3, 'Person_sitting': 4, 'Cyclist': 5, 'Tram': 6, 'Misc': 7, 'DontCare': 8}.

Will the model weights be updated according to my class mapping or it'll update it's weights according to the class labels to which it was pretrained?

I mean to ask
let's say I mapped car as 0, but while pretraining you used car as 5, what will model predit after I train it on custom dataset?

**Reply by pderrenger** [reply]
Hi there!

Great question! When you train your model on a custom dataset with your own class mapping, the model will update its weights according to your new class labels. So, if you map 'Car' as 0, the model will learn to predict 'Car' as class 0, regardless of the class ID used during pretraining. 

Here's a quick example to ensure your training setup:
```python
from ultralytics import YOLO

# Load a pretrained model
model = YOLO("yolov8n.pt")

# Train the model with your custom dataset
results = model.train(data="your_dataset.yaml", epochs=100, imgsz=640)
```

Happy training! 🚀

**Comment by skp4595** [comment]
I'd like to contact you via linkedin or mail @Ultralytics team.
I have a few more questions w.r.to the calibration. I tried the pretrained model and fine tuned model on kitti dataset, I'd like to discuss it over linkedin or mail.

**Comment by subbubhat99** [comment]
Hey there!
I'm trying to train RT-DETR-L on my custom dataset containing two classes of objects i.e.  0: 'Chewing Gum' and 1: 'Cigarette Butt'.
If I structure the dataset slightly differently i.e. having train, val and test folders with each of them having a data and labels sub-folders, why are my labels not being found by the code?
The YAML file contents look like: 

path: path\\to\\root\\directory  #Path to the root directory of dataset
train: .\train\data #Path to the fraction of all images dedicated to training
val: .\val\data #Path to the fraction of all images dedicated to validation
test: .\test\data #Path to the fraction of all images dedicated to testing
train_labels: .\train\labels #Path to the fraction of all annotations dedicated to training
val_labels: .\val\labels #Path to the fraction of all annotations dedicated to validation
test_labels: .\test\labels #Path to the fraction of all annotations dedicated to testing
nc : 2 #Number of classes in the dataset
PS: I used absolute path notations for the root directory.

#Classes
names: ['Cigarette Butt', 'Chewing Gum']
Look forward to your support in this regard

**Reply by glenn-jocher** [reply]
Hello!

It looks like the issue might be due to the YAML file structure. In the current Ultralytics YOLO format, the paths for labels are not separately defined in the YAML file. Instead, the labels should be in the same directory as the images they correspond to, and the directory structure should be specified under the `train`, `val`, and `test` keys directly.

Here’s how you can adjust your YAML configuration:

```yaml
path: path\\to\\root\\directory  # Path to the root directory of the dataset
train: train\\data  # Path to training images and labels
val: val\\data  # Path to validation images and labels
test: test\\data  # Path to testing images and labels
nc: 2  # Number of classes
names: ['Cigarette Butt', 'Chewing Gum']
```

Make sure that each image in `train\\data`, `val\\data`, and `test\\data` has a corresponding label file in the same directory. The label files should be named identically to the images, but with a `.txt` extension.

If your labels are still not being found, ensure that the paths are correctly specified and accessible from the directory where you are running your training script.

Hope this helps! Let us know if you have any more questions. 😊

**Reply by subbubhat99** [reply]
Hey Glenn!
Thanks for taking the time out to clarify that! I did end up changing the structure of my dataset to have two folders images and labels, each of which now have train, test and val sub-folders(similar to the format shown for coco.yaml). And that has worked.
Kind regards,
Subramanya

**Reply by glenn-jocher** [reply]
@subbubhat99 hello Subramanya,

Great to hear that adjusting your dataset structure resolved the issue! If you have any more questions or need further assistance as you continue with your project, feel free to reach out. Happy training! 😊

Kind

**Reply by subbubhat99** [reply]
Hey again,
As the structuring of the dataset file changed and the model I was trying to train(RT-DETR) successfully completed training on the custom data and runs for inference too, the next step in my project is to integrate this object detector with a tracker such as ByteTrack or BoT-SORT to facilitate simultaneous object detection and tracking of targets in the scene(in my case its two types of trash i.e. Cigarettes and chewing gums lying on the street). While the github repo explains how to make YOLO+ByteTrack/BoT-SORT work, I was wondering if a similar process could be followed to make RT-DETR+ByteTrack work as well?
Look forward to hearing from you regarding this.
Thanks!

**Reply by glenn-jocher** [reply]
Hello!

It's great to hear that your model training on the custom dataset was successful and that you're now looking to integrate object tracking. While the integration of RT-DETR with ByteTrack or BoT-SORT isn't directly documented like YOLO models, the process should be conceptually similar.

You would need to ensure that the output format of RT-DETR (detections including bounding boxes and confidence scores) is compatible with the input requirements of the tracking algorithms. Typically, this involves ensuring that the detection model outputs are formatted correctly for the tracker to process (e.g., converting to the appropriate coordinate system and ensuring confidence scores are passed).

If RT-DETR outputs are compatible, you can then proceed by feeding these detections into ByteTrack or BoT-SORT as you would with detections from a YOLO model. You might need to adjust some parameters or settings in the tracking configuration to optimize performance for your specific use case.

If you encounter any specific issues or have more detailed questions as you proceed, feel free to reach out!

Best of luck with your project, and looking forward to hearing about your progress with integrating detection and tracking!

**Comment by rahul97jp** [comment]
Hi, I am using the below code to cnvert coco-text json annaotation to YOLO format. It gives output "COCO data converted successfully.
Results saved to /content/coco_converted".

But the folder is empty.

from ultralytics.data.converter import convert_coco
convert_coco(labels_dir="/content/drive/MyDrive/Colab/coco/cocotext.v2.json")

**Reply by glenn-jocher** [reply]
Hello,

It seems like the conversion process completed without errors, but the output folder being empty suggests the `labels_dir` path might not be correctly pointing to your annotations. Ensure that the path `/content/drive/MyDrive/Colab/coco/cocotext.v2.json` correctly points to your COCO-format annotations file. Also, verify that the JSON file contains the expected data in the correct format.

If the path and file are correct, it could be helpful to check the permissions of the directory to ensure that your script has write access to save the converted files.

If you continue to face issues, please provide more details about the contents and structure of your JSON file, and I'll be glad to assist further!

**Reply by rahul97jp** [reply]
The directory has the write access.
The structure of the json is (sample of it): 
{"cats":{},"anns":{"45346":{"mask":[468.9,286.7,468.9,295.2,493.0,295.8,493.0,287.2],"class":"machine printed","bbox":[468.9,286.7,24.1,9.1],"image_id":217925,"id":45346,"language":"english","area":206.06,"utf8_string":"New","legibility":"legible"},"153036":{"mask":[344.5,261.5,348.1,261.5,348.2,263.4,344.5,263.4],"class":"machine printed","bbox":[344.5,261.5,3.7,1.9],"image_id":483569,"id":153036,"language":"english","area":6.93,"utf8_string":"","legibility":"illegible"},"125303":{"mask":[362.4,280.9,359.2,286.3,367.1,291.0,369.9,285.0],"class":"machine printed","bbox":

https://bgshih.github.io/cocotext/

**Reply by glenn-jocher** [reply]
Hello,

Thank you for sharing the structure of your JSON file. It appears that the JSON format you provided is not directly compatible with the expected COCO format that the `convert_coco` function is designed to handle. The function expects the standard COCO format with fields like "images", "annotations", and "categories", which are not present in your JSON structure.

To resolve this, you'll need to either adjust your JSON to match the COCO format or write a custom script to convert your specific JSON structure to the YOLO format. If you need further assistance with writing a custom conversion script, please let me know, and I'd be happy to help!

**Comment by subbubhat99** [comment]
Hi Glenn,
Thanks for taking the time out to provide those tips. You're right about
bridging the gap between the format of RT-DETR outputs and Byte-Track's
expected inputs. If I understood correctly, RT-DETR has a different bbox
coordinate system than YOLO(which has cxcywh) right?
Need to be sure about this before proceeding to integrate the two.
Thanks,
Kind regards,
Subramanya


On Fri, May 31, 2024 at 3:57 PM Glenn Jocher ***@***.***>
wrote:

> Hello!
>
> It's great to hear that your model training on the custom dataset was
> successful and that you're now looking to integrate object tracking. While
> the integration of RT-DETR with ByteTrack or BoT-SORT isn't directly
> documented like YOLO models, the process should be conceptually similar.
>
> You would need to ensure that the output format of RT-DETR (detections
> including bounding boxes and confidence scores) is compatible with the
> input requirements of the tracking algorithms. Typically, this involves
> ensuring that the detection model outputs are formatted correctly for the
> tracker to process (e.g., converting to the appropriate coordinate system
> and ensuring confidence scores are passed).
>
> If RT-DETR outputs are compatible, you can then proceed by feeding these
> detections into ByteTrack or BoT-SORT as you would with detections from a
> YOLO model. You might need to adjust some parameters or settings in the
> tracking configuration to optimize performance for your specific use case.
>
> If you encounter any specific issues or have more detailed questions as
> you proceed, feel free to reach out!
>
> Best of luck with your project, and looking forward to hearing about your
> progress with integrating detection and tracking!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/ultralytics/discussions/12639#discussioncomment-9619121>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWCYDT4LZQEZZ25OALL2U3ZFB6UBAVCNFSM6AAAAABHTL2AXWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TMMJZGEZDC>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Subramanya Bhat
5/16, University Road
University of Delhi
Delhi-110 007
Phone:27662141

**Reply by glenn-jocher** [reply]
@subbubhat99 hi Subramanya,

Yes, you're correct. RT-DETR typically outputs bounding boxes in a different format compared to the YOLO models, which use the cxcywh format (center x, center y, width, height). Before integrating RT-DETR outputs with Byte-Track, you'll need to convert these bounding box coordinates into the expected format.

If you need further assistance with the conversion process or any other integration details, feel free to reach out!

Kind

**Comment by wizard0696** [comment]
Hello. I'm new to YOLO, so I need help with a project because I don't know how to select the classes I want, such as vehicle classes. I used CLI and defined it as follows:
yolo predict model=yolov8s.pt source=bus.jpg classes= 1 2 3 4 5 6 7 8

**Reply by wizard0696** [reply]
Solved. It had to be just like this, without spaces:
yolo predict model=yolov8s.pt source=bus.jpg classes=[1,2,3,4,5,6,7,8]

**Reply by glenn-jocher** [reply]
@wizard0696 hello! To select specific classes such as vehicle classes using the CLI, you should list the class indices without spaces. For example, you can use the following command:

```bash
yolo predict model=yolov8s.pt source=bus.jpg classes=1,2,3,4,5,6,7,8
```

This will filter the predictions to only include the specified classes. If you need further details, please refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/detect/).

**Comment by robin-bhooman** [comment]
Hello,

I am looking to prepare a dataset for multi-label detection, where each object in an image could belong to multiple classes. I understand that the .txt file for labels have the following format for each object in an image: `class xc yc w h`. To suit this for a multi-label case, would it suffice to simply add duplicate bounding boxes with different classes? For example, given this [image](https://www.researchgate.net/profile/Maria-Andersson-9/publication/275084997/figure/fig4/AS:667642817966085@1536189860592/Results-from-object-detection-and-tracking-The-car-and-the-people-framed-in-green-are.jpg) with three objects, I want the car to have the labels [vehicle, suv], and the two people to have the labels [person, male]. Can I make the label.txt for this image look like this, and would this work to train a multi-label detection model?
```txt
cls_1 xc_1 yc_1 w_1 h_1
cls_2 xc_1 yc_1 w_1 h_1 
cls_3 xc_2 yc_2 w_2 h_2
cls_4 xc_2 yc_2 w_2 h_2
cls_3 xc_3 yc_3 w_3 h_3
cls_4 xc_3 yc_3 w_3 h_3  
```

**Reply by glenn-jocher** [reply]
@bhoo-git yes, you can add duplicate bounding boxes with different classes for multi-label detection. Your example label file format is correct and will work for training a multi-label detection model. For more details, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/detect/).

**Comment by ALIABB17** [comment]
Hello,
I am using multiple networks to train my dataset, one of them is ultralytics yolov8, all of them use mIoU to evaluate the performance except yolov8, does it have a built-in comutation for mIoU? if not how can i impement it?
Thank you in advance.

**Reply by glenn-jocher** [reply]
Ultralytics YOLOv8 does not have a built-in mIoU computation. You can implement it by calculating the intersection over union for each class and averaging them. For more details on dataset handling, visit our documentation.

**Comment by wgdcxy** [comment]
WARNING ⚠️ imgsz=[2448] must be multiple of max stride 32, updating to [2464]
train: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttb... 0 images, 30 backgrounds, 0 corrupt: 100%|██████████| 30/30 [00:00<00:00, 1138.41it/s]
train: WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttb.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
train: WARNING ⚠️ Cache directory D:\Textile_testing\ultralytics-seg\cs\labels is not writeable, cache not saved.
WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttb.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttj... 0 images, 30 backgrounds, 0 corrupt: 100%|██████████| 30/30 [00:00<00:00, 1617.94it/s]
val: WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: WARNING ⚠️ Cache directory D:\Textile_testing\ultralytics-seg\cs\labels is not writeable, cache not saved.
WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
Plotting labels to runs\segment\train3\labels.jpg... 

这个问题该如何解决？

**Reply by glenn-jocher** [reply]
It looks like your dataset labels are missing or not formatted correctly. Please ensure your labels are in the YOLO format and check the directory permissions. For detailed guidance, visit the dataset formatting section at https://docs.ultralytics.com/datasets/detect/.

**Comment by wgdcxy** [comment]
Hello, my label and image files are stored according to the requirements, but they are still not recognized, and the following is an error message
Thank you in advance



WARNING ⚠️ imgsz=[2448] must be multiple of max stride 32, updating to [2464]
train: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttb... 0 images, 30 backgrounds, 0 corrupt: 100%|██████████| 30/30 [00:00<00:00, 1138.41it/s]
train: WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttb.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
train: WARNING ⚠️ Cache directory D:\Textile_testing\ultralytics-seg\cs\labels is not writeable, cache not saved.
WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttb.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttj... 0 images, 30 backgrounds, 0 corrupt: 100%|██████████| 30/30 [00:00<00:00, 1617.94it/s]
val: WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: WARNING ⚠️ Cache directory D:\Textile_testing\ultralytics-seg\cs\labels is not writeable, cache not saved.
WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
Plotting labels to runs\segment\train3\labels.jpg...

**Reply by glenn-jocher** [reply]
It seems like your dataset labels are not being recognized. Please ensure your label files are in the correct YOLO format and that the directory is writable. For detailed guidance, visit the dataset formatting section at https://docs.ultralytics.com/datasets/detect/.

**Reply by wgdcxy** [reply]
Hi, but my file path is very accurate, and I checked it N times, so is there any otherreason? Thank you

**Reply by glenn-jocher** [reply]
@wgdcxy please ensure your label files are in the correct YOLO format, with one `.txt` file per image containing normalized coordinates. Also, verify that the directory permissions allow writing. For more details, refer to the dataset formatting guide linked earlier.

**Reply by wgdcxy** [reply]
Hello, I'm a student whose English is not particularly good, so some places may not be clear, and secondly, this is not an error but a warning.I don't know why it's the background that reads, not the images.
The warnings are as follows：
val: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttj... 0 images, 30 backgrounds, 0 corrupt: 100%| 30/30 [00:00<00:00, 495.76it/s]
val: WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: WARNING ⚠️ Cache directory D:\Textile_testing\ultralytics-seg\cs\labels is not writeable, cache not saved.
WARNING ⚠️ No labels found in D:\Textile_testing\ultralytics-seg\cs\labels\ttj.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
Plotting labels to runs\segment\train4\labels.jpg... 

Then this warning tells me that the cache file is missing, and I don't know why I can't generate a cache file for the file I created, so I'm here to ask you guys.
Below is my own yaml file：
    # Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
    path: ../cs  # dataset root dir
    train: images/ttb  # train images (relative to 'path') 128 images
    val: images/ttj # val images (relative to 'path') 128 images
    test:  # test images (optional)
    # Classes
    names :
            0 : wf

Below is my label file：
0 0.6799428104575164 0.5005859375 0.6844771241830065 0.500537109375 0.6845996732026144 0.497265625 0.6802696078431373 0.497119140625
0 0.6803104575163399 0.4935546875 0.6848039215686275 0.493505859375 0.6847630718954248 0.497119140625 0.6802696078431373 0.49697265625
0 0.6802696078431373 0.49013671875 0.6849673202614379 0.490185546875 0.6849264705882353 0.49345703125 0.6803513071895425 0.493359375
0 0.6805555555555556 0.48662109375 0.6850898692810458 0.486767578125 0.6850490196078431 0.490234375 0.6803513071895425 0.490234375
0 0.6808415032679739 0.483447265625 0.6852941176470588 0.48349609375 0.6852124183006536 0.486767578125 0.6806781045751634 0.486669921875
0 0.6806372549019607 0.47998046875 0.6854166666666666 0.47998046875 0.6853349673202613 0.48349609375 0.6808415032679738 0.48359375


Finally, this yolo can be run, but I don't know why the result is 0.
Here's how it works：
Transferred 381/417 items from pretrained weights
      195/200      8.73G          0          0        nan          0          0       2464: 100%|██████████| 8/8 [01:37<00:00, 12.13s/it]
       Class     Images  Instances   Box(P    R   mAP50  mAP50-95)   Mask(P    R    mAP50  mAP50-95): 100%|[00:31<00:00,  7.91s/it]
          all         30          0              0          0          0          0                  0          0          0          0
Finally, thank you for explaining it to me

**Reply by wgdcxy** [reply]
By the way, the size of my picture is 2448x2048, so the imgsz entered is 2448, and your code changed the size of my picture to 2464, is it related to this.
imgsz=[2448] must be multiple of max stride 32, updating to [2464]

**Reply by wgdcxy** [reply]
I apologize, I found a small mistake of mine. I wrote the file name as "labels" instead of "lables" which I intended to write.Please forgive my carelessness.After all, who doesn't make mistakes. I'll come back to you next time I have a problem. Thank you for your help.

**Reply by wgdcxy** [reply]
Transferred 417/417 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
WARNING ⚠️ imgsz=[2448] must be multiple of max stride 32, updating to [2464]
train: Scanning D:\Textile_testing\ultralytics-seg\cs\labels\ttb.cache... 30 images, 0 backgrounds, 0 corrupt: 100%|██████████| 30/30 [00:00<?, ?it/s]
Hello, here I am again, with a new problem, how to solve this problem about image size?Ask for answers, thank you

**Reply by glenn-jocher** [reply]
@wgdcxy to address the image size issue, ensure your `imgsz` is a multiple of 32, as required by the model's architecture. This adjustment helps maintain compatibility with the model's stride. If you have further questions, feel free to ask!

**Reply by wgdcxy** [reply]
WARNING ⚠️ NMS time limit 0.900s exceeded.
Hello, my teacher who has never met, what should I do with this warning?

**Reply by glenn-jocher** [reply]
@wgdcxy to address the NMS time limit warning, try increasing the `nms` parameter in your configuration to allow more processing time. If you have further questions, feel free to ask!

**Reply by wgdcxy** [reply]
Hello, I'm here again. The issue this time is still the same warning. I adjusted the max_time_img to 0.1 in ops. py, but I don't feel it's right. So I came to ask for your help because it still shows this warning. I hope you can tell me where to make the adjustment. Thank you again.
WARNING ⚠️ NMS time limit 0.900s exceeded

**Reply by wgdcxy** [reply]
How can these values be improved? We can only continue training or increase the training dataset.
      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
     11/500        11G        nan       6.21        nan        nan        132       2464: 100%|██████████| 8/8 [01:56<00:00, 14.59s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:54<00:00, 13.56s/it]
                   all         30        664          0          0          0          0          0          0          0          0

**Reply by glenn-jocher** [reply]
To improve your model's performance, consider increasing the dataset size, enhancing data quality, or adjusting hyperparameters. If issues persist, reviewing the dataset for label accuracy and distribution might help.

**Comment by milanaveed** [comment]
Hello, can you please fix the broken images on this page? Can't see the label file example.

**Reply by glenn-jocher** [reply]
@milanaveed thank you for bringing this to our attention. We'll look into the issue with the images on the page. In the meantime, you can refer to the text description for guidance on the label file format.

**Comment by Yuxuan1998** [comment]
Hi team, I am a little bit confused with the dataset folder structure.
For most dataset I used, it strunctured in the way that each of `train/` and `val/` folders contains `images/` and `labels/` folders.
However, the COCO8 dataset example shows the opposite way, each of `images/` and `labels/` folder contains `test/` and `val/` folders.
Does the folder structure matter here? Thanks in advance

**Reply by glenn-jocher** [reply]
@Yuxuan1998 the folder structure does not matter as long as the paths are correctly defined in your dataset's YAML file. Both structures are supported, and you can specify the relative paths to the `images/` and `labels/` directories accordingly. For details, refer to the [Object Detection Datasets Overview](https://docs.ultralytics.com/datasets/detect/). Let us know if you need further clarification!

**Comment by astronitya** [comment]
Hello Team,
 I am using YOLO for detecting trails of slow-moving astronomical objects in my images. While the model detects the trails, the bounding boxes predicted around the objects are not tight enough. The boxes tend to be too loose or not well-shaped to fit the elongated, irregular trails, resulting in inaccurate object localization. I suspect, YOLO's standard rectangular bounding box not being well-suited for the elongated, curving shapes of the trails. Could you provide guidance on how to improve bounding box accuracy for this specific use case?

**Reply by glenn-jocher** [reply]
@astronitya thank you for your question! For detecting elongated or irregular trails, consider using Oriented Bounding Boxes (OBB), which may better fit your objects' shapes compared to standard rectangular boxes. You can format your dataset in the YOLO OBB format and train a YOLO11 OBB model. Please refer to the [Oriented Bounding Box (OBB) Datasets Overview](https://docs.ultralytics.com/datasets/obb/) for details on dataset formatting and training. Let me know if you need further clarification!

**Reply by astronitya** [reply]
@glenn-jocher Thankyou very much for the prompt response. I am going to work on your suggestions.

**Reply by glenn-jocher** [reply]
@astronitya you're welcome! We're glad the suggestions were helpful. Feel free to share your progress or reach out with further questions. The Ultralytics community is always here to support you. Good luck with your project!

**Comment by bimal-gajera** [comment]
Hello,
I'm working on multiple dataset formats for object detection. For the YOLO dataset format, is there any way I can rename the labels directory and put it in yaml file?
Something like this:
```
path: path_to_root_directory
train: train\images
val: val\images
test: test\images

labels_dir: yolo

nc: 1
names: ['object']
```

yolo directory has yolo annotations (*.txt files), and it's inside respective train, val and test directory.

**Reply by glenn-jocher** [reply]
@bimal-gajera currently, the `labels_dir` key is not supported in the YAML file configuration for YOLO datasets. Label paths must follow the standard directory structure where label files are placed alongside their corresponding images in subdirectories like `train`, `val`, or `test`. For more details, refer to the [Object Detection Datasets Overview](https://docs.ultralytics.com/datasets/detect/).

**Comment by Fatima-khanjar** [comment]
Hello,
I hope this message finds you well.

I have a question regarding fine-tuning a pre-trained YOLO model (e.g., trained on the COCO dataset) on my custom dataset. My dataset includes both new classes (not present in COCO) and the same existing classes (e.g., toothbrush, car), but with different layouts. Specifically, I would like to add my new classes starting from index 80, while still leveraging the pre-trained weights for the common classes, which are the same but with different layouts.

Is it possible to use the pre-trained weights for the common classes and only train the model on the new classes, while ensuring that the pre-trained knowledge for the common classes is preserved? If so, could you please guide me on how to proceed?

Thank you for your time and assistance.
best regards,
